cucumber-core/src/main/java/io/cucumber/core/plugin/JUnitFormatter2.java

```diff
@@ -0,0 +1,350 @@
+            return testCaseStartedIdToResult.values().stream()
+                    .map(TestStepResult::getStatus)
+                    .collect(groupingBy(identity(), counting()));
+            // @formatter:on
+        }
+
+        int getTestCaseCount() {
+            return testCaseStartedIdToStartedInstant.size();
+        }
+
+        String getPickleName(String testCaseStartedId) {
+            String testCaseId = testCaseStartedIdToTestCaseId.get(testCaseStartedId);
+            String pickleId = testCaseIdToPickleId.get(testCaseId);
+            return pickleIdToPickleName.get(pickleId);
+        }
+
+        public String getFeatureName(String testCaseStartedId) {
+            String testCaseId = testCaseStartedIdToTestCaseId.get(testCaseStartedId);
+            String pickleId = testCaseIdToPickleId.get(testCaseId);
+            String astNodeId = pickleIdToScenarioAstNodeId.get(pickleId);
+            return scenarioAstNodeIdToFeatureName.get(astNodeId);
+        }
+
+        public Deque<String> testCaseStartedIds() {
+            return testCaseStartedIds;
+        }
+
+        private static final io.cucumber.messages.types.Duration ZERO_DURATION = new io.cucumber.messages.types.Duration(
+            0L, 0L);
+        private static final TestStepResult SCENARIO_WITH_NO_STEPS = new TestStepResult(ZERO_DURATION, null, PASSED); // By
+                                                                                                                      // definition,
+                                                                                                                      // but
+                                                                                                                      // see
+                                                                                                                      // https://github.com/cucumber/gherkin/issues/11
+        public TestStepResult getTestCaseStatus(String testCaseStartedId) {
+            return testCaseStartedIdToResult.getOrDefault(testCaseStartedId, SCENARIO_WITH_NO_STEPS);
+        }
+
+    }
+
+    private static class XmlReportWriter {
+        private final NumberFormat numberFormat = NumberFormat.getInstance(Locale.US);
+        private final XmlReportData data;
+
+        public XmlReportWriter(XmlReportData data) {
+            this.data = data;
+        }
+
+        private void writeXmlReport(Writer out) throws XMLStreamException {
+            XMLOutputFactory factory = XMLOutputFactory.newInstance();
+            XMLStreamWriter writer = factory.createXMLStreamWriter(out);
+            writer.writeStartDocument("UTF-8", "1.0");
+            newLine(writer);
+            writeTestsuite(data, writer);
+            writer.writeEndDocument();
+            writer.flush();
+        }
+
+        private void writeTestsuite(XmlReportData data, XMLStreamWriter writer) throws XMLStreamException {
+            writer.writeStartElement("testsuite");
```

Comment:
Consider adding a `testsuites` element so the `testsuite` element can match the feature file.

E.g: https://github.com/cucumber/cucumber-js/pull/2121/files#diff-7859e8a3e68f4017ba6bfd6f7ee909dd52bc87c2ddc84913259c446198ea9f36R100-R109